### PR TITLE
HackStudio: Fix issues with retaining state of previous project after opening a different one

### PR DESCRIPTION
**Pull request by donaghylennon**
Previously when a you opened a different project or created a new one, many of the widgets within HackStudio would not update to reflect this. This pull request fixes many of these issues, as well as fixing a bug with GitWidget in particular, which would cause HackStudio to crash when trying to create a git repository after switching to a different project.

### CODE REVIEW

**Comment by ADKaster** [comment]
Haven't looked at these changes but, what is the behavior when you have unsaved changes and you open a different project?

I think the desired behavior would be it prompts you to  save the file(s) with a dialog, and if you cancel we don't open the new project.

**Comment by donaghylennon** [comment]
It currently just silently opens the new project without warning, it also does this when you close HackStudio. I just checked and there is some code to do this, it's used when you have unsaved changes and try to build. It looks like it'd be pretty easy to use that for both these cases, would it make sense to include them in this pull request?

**Comment by ADKaster** [comment]
If you wanna add it then go for it, otherwise we can open an issue to track it for another PR

**Comment by donaghylennon** [comment]
I did a bit of digging and it turns out it's supposed to warn if you have unsaved changes before closing HackStudio as well as before building, but for some reason this code is never called, it seems like this got broken at some point (before my changes). Looks like this will need to be fixed before I try to use it myself, I'll try to get to the bottom of it.

**Comment by donaghylennon** [comment]
@ADKaster I figured out what it was, the editor's on_change callback was being overwritten so that the GML preview could be updated every time you changed something. The best way I could see to fix this right now was to make it update the preview when you save instead, but I'm sure this could be improved on, maybe doing something similar to how the ToDo entries are updated.

File: Userland/DevTools/HackStudio/HackStudioWidget.cpp [hunk 1]
```diff
@@ -300,6 +300,33 @@ bool HackStudioWidget::open_file(const String& full_filename)
     return true;
 }
 
+void HackStudioWidget::close_file(const String& filename)
+{
+    m_open_files.remove(filename);
+    m_open_files_vector.remove_all_matching(
+        [&filename](const String& element) { return element == filename; });
```
Comment (itamar8910) [inline Userland/DevTools/HackStudio/HackStudioWidget.cpp:308]: Please use east const

File: Userland/DevTools/HackStudio/HackStudioWidget.cpp [hunk 2]
```diff
@@ -300,6 +300,33 @@ bool HackStudioWidget::open_file(const String& full_filename)
     return true;
 }
 
+void HackStudioWidget::close_file(const String& filename)
```
Comment (itamar8910) [inline Userland/DevTools/HackStudio/HackStudioWidget.cpp:304]: Perhaps `close_file_in_all_editors` would be a better name?

File: Userland/DevTools/HackStudio/Git/GitWidget.h [hunk 3]
```diff
@@ -24,6 +24,7 @@ class GitWidget final : public GUI::Widget {
     void refresh();
     void set_view_diff_callback(ViewDiffCallback callback);
     bool initialized() const { return !m_git_repo.is_null(); };
+    void change_repo(const LexicalPath& repo_root);
```
Comment (itamar8910) [inline Userland/DevTools/HackStudio/Git/GitWidget.h:27]: Please use east const

File: Userland/DevTools/HackStudio/HackStudioWidget.cpp [hunk 4]
```diff
@@ -294,7 +294,6 @@ bool HackStudioWidget::open_file(const String& full_filename)
     current_editor().set_focus(true);
 
     current_editor().on_cursor_change = [this] { update_statusbar(); };
-    current_editor().on_change = [this] { update_gml_preview(); };
```
Comment (itamar8910) [inline Userland/DevTools/HackStudio/HackStudioWidget.cpp:297]: IMO updating the GML preview on every change is a feature we want to have.

How about changing the API of the `EditorWrapper` and allow registering a callback function
that would get called on every editor change?
Then `EditorWrapper` could still internally take care of marking the document as dirty, and `update_gml_preview()` can get called regardless.

**Review by itamar8910 — CHANGES_REQUESTED** [review]
These are some really solid improvements :+1: 
Please see my comments.

**Comment by donaghylennon** [comment]
@itamar8910 I made the changes you asked for, now the GML preview updates on every change and the document is correctly marked as dirty :)

**Review by itamar8910 — APPROVED** [review]
LGTM :)

**Pull request merged**